esp32: Remove support for ESP-IDF <5.2

projectgus opened the pull request:
### Summary

When we merged support for ESP-IDF V5.2.2 in #15523, we discussed dropping support for older versions after the new version was bedded in. ([Related Discord discussion here](https://discord.com/channels/574275045187125269/952412628926472273/1265452441932136550)). This is a draft for that PR!

In two commits, one to remove all the #ifdef-ed code and one that reverts 27279e69b4 from #13775 (adding a 5.2-specific sdkconfig variable).

### Testing

* Haven't done a lot of on-device testing, but built ports for ESP32_GENERIC, ESP32_GENERIC_C3, ESP32_GENERIC_S3, ESP32_GENERIC_C6 and verified the binary size was the same for both master and this branch (ESP-IDF V5.2.2).

### Trade-offs and Alternatives

* Was originally going to keep the `SDKCONFIG_IDF_VERSION_SPECIFIC` variable from #13775 in case we need it in the future, but I reckon it's better to revert it here and we can always un-revert it later if a need arises for it.

projectgus committed:
esp32: Remove IDF-version-specific sdkconfig.

This reverts commit 27279e69b4d74b255bd8d8c5dd6668c97c942776
(plus removes some additional references to the
SDKCONFIG_IDF_VERSION_SPECIFIC CMake variable.)

Relevant sdkconfig options are added into sdkconfig.base now
that IDF >=5.2.0 is required.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
esp32: Drop support for ESP-IDF below V5.2.0.

Specifically, remove all conditional compilation for these earlier versions
and change the idf_component.yml specifiers to require >=5.2.0.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
Sorry andrew and damien, I thought I'd cancelled the review request before opening (as realised it's still a draft).

dpgeorge commented:
You (and others) did a lot of testing of IDF 5.2.2 in #15523, so I'm happy to move forward with this as the minimum supported version.  Also, firmware size was reduced a lot, so that means users won't struggle to upgrade if they are short on flash.

Since we will make a v1.24.1 (and possibly further patch releases on that v1.24 branch), we can have v1.24 as the last MicroPython version to support IDF < 5.2.2, and users can stick to that release branch if necessary.

projectgus commented on ports/esp32/boards/sdkconfig.base:
```diff
@@ -50,6 +50,8 @@ CONFIG_ESP_SYSTEM_MEMPROT_FEATURE=n
 CONFIG_FREERTOS_THREAD_LOCAL_STORAGE_POINTERS=2
 CONFIG_FREERTOS_SUPPORT_STATIC_ALLOCATION=y
 CONFIG_FREERTOS_ENABLE_STATIC_TASK_CLEAN_UP=y
+CONFIG_FREERTOS_ENABLE_STATIC_TASK_CLEAN_UP=n
```
Oops!

dpgeorge commented on ports/esp32/README.md:
```diff
@@ -28,7 +28,7 @@ manage the ESP32 microcontroller, as well as a way to manage the required
 
 The ESP-IDF changes quickly and MicroPython only supports certain versions.
-Currently MicroPython supports v5.0.4, v5.0.5, v5.1.2, v5.2.0, v5.2.2.
+Currently MicroPython supports v5.2.0 and v5.2.2.
```
The tag is actually `v5.2`, see #16195.

projectgus commented:
Rebased now the linked PRs are merged, this PR now includes a revert of #16127. Re-ran unit tests on S3 and verified Wi-Fi is working on C3.

dpgeorge commented on ports/esp32/main_esp32/idf_component.yml:
```diff
@@ -2,4 +2,4 @@
   espressif/mdns: "~1.1.0"
   idf:
-    version: ">=5.0.4"
+    version: ">=5.2.0"
```
Is this correct even though the IDF tag is `v5.2`?

dpgeorge replied:
I'm pretty sure it's correct, just wanted to double check.

projectgus replied:
It is, these are SemVer rather than git tags: https://docs.espressif.com/projects/idf-component-manager/en/latest/reference/versioning.html#version-precedence

dpgeorge commented on ports/esp32/mpthreadport.c:
```diff
@@ -41,7 +41,7 @@
 #define MP_THREAD_PRIORITY                              (ESP_TASK_PRIO_MIN + 1)
 
-#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 2, 0) && !CONFIG_FREERTOS_ENABLE_STATIC_TASK_CLEAN_UP
+#if !CONFIG_FREERTOS_ENABLE_STATIC_TASK_CLEAN_UP
```
Is this if logic needed at all now that this option is always disabled?

projectgus replied:
Oh yeah, have pulled all this out!

dpgeorge commented on ports/esp32/network_wlan.c:
```diff
@@ -760,10 +756,8 @@ static MP_DEFINE_CONST_DICT(wlan_if_locals_dict, wlan_if_locals_dict_table);
 
 #if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 2, 0)
 _Static_assert(WIFI_AUTH_MAX == 13, "Synchronize WIFI_AUTH_XXX constants with the ESP-IDF. Look at esp-idf/components/esp_wifi/include/esp_wifi_types.h");
-#elif ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 0, 5) && ESP_IDF_VERSION < ESP_IDF_VERSION_VAL(5, 1, 0) || ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 1, 2)
```
I think there's some if logic above for the new SEC constants that can be removed.

projectgus replied:
Thanks, have removed.

projectgus replied:
Thanks, should be all cleaned out now.

projectgus committed:
esp32: Simplify thread cleanup.

Now we only support the case of
!CONFIG_FREERTOS_ENABLE_STATIC_TASK_CLEAN_UP, can simplify
the cleanup code.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
Have rebased onto the v5.3 changes, ran one lot of unit tests on C6 and also re-ran the before/after binary size comparison on the boards mentioned in the PR description.

dpgeorge commented:
> ran one lot of unit tests on C6 and also re-ran the before/after binary size comparison on the boards mentioned in the PR description.

Great, thanks for re-testing!

Now merged.
